Tools/unicode/dawg.py

```diff
@@ -0,0 +1,486 @@
+# Original Algorithm:
+# By Steve Hanov, 2011. Released to the public domain.
+# Please see http://stevehanov.ca/blog/index.php?id=115 for the accompanying article.
+#
+# Adapted for PyPy/CPython by Carl Friedrich Bolz-Tereick
+#
+# Based on Daciuk, Jan, et al. "Incremental construction of minimal acyclic finite-state automata."
+# Computational linguistics 26.1 (2000): 3-16.
+#
+# Updated 2014 to use DAWG as a mapping; see
+# Kowaltowski, T.; CL. Lucchesi (1993), "Applications of finite automata representing large vocabularies",
+# Software-Practice and Experience 1993
+
+from collections import defaultdict
+from functools import cached_property
+
+DEBUG = False
+
+
+# This class represents a node in the directed acyclic word graph (DAWG). It
+# has a list of edges to other nodes. It has functions for testing whether it
+# is equivalent to another node. Nodes are equivalent if they have identical
+# edges, and each identical edge leads to identical states. The __hash__ and
+# __eq__ functions allow it to be used as a key in a python dictionary.
+
+
+class DawgNode:
+
+    def __init__(self, dawg):
+        self.id = dawg.next_id
+        dawg.next_id += 1
+        self.final = False
+        self.edges = {}
+
+        self.linear_edges = None # later: list of (string, next_state)
+
+    def __str__(self):
+        if self.final:
+            arr = ["1"]
+        else:
+            arr = ["0"]
+
+        for (label, node) in sorted(self.edges.items()):
+            arr.append(label)
+            arr.append(str(node.id))
+
+        return "_".join(arr)
+    __repr__ = __str__
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+    @cached_property
+    def num_reachable_linear(self):
+        # returns the number of different paths to final nodes reachable from
+        # this one
+
+        count = 0
+        # staying at self counts as a path if self is final
+        if self.final:
+            count += 1
+        for label, node in self.linear_edges:
+            count += node.num_reachable_linear
+
+        return count
+
+
+class Dawg:
+    def __init__(self):
+        self.previous_word = ""
+        self.next_id = 0
+        self.root = DawgNode(self)
+
+        # Here is a list of nodes that have not been checked for duplication.
+        self.unchecked_nodes = []
+
+        # Here is a list of unique nodes that have been checked for
+        # duplication.
+        self.minimized_nodes = {}
+
+        # Here is the data associated with all the nodes
```

Comment:
"All the nodes" sounds to me like the keys should be nodes. Suggest something like this instead:
```suggestion
        # word --> value and value --> word mappings
```